### PR TITLE
[FIX] Fix heredoc bad substitution

### PR DESCRIPTION
**Pull request by itislu**
Now all heredoc functions check the return value of `ft_expander()` and only report `HEREDOC_ERROR` if a `MALLOC_ERROR` occcured in the expander. This previously caused the shell to exit in case of a bad substitution.

Also we forgot to increase `cur_len` when using `ft_snprintf()`. This caused every new line in the heredoc content to be overwritten with the delimiter (`\n`).
I fixed it by using the return value of `ft_snprintf()` to increase the string index (`cur_len`).

### CODE REVIEW

**Comment by itislu** [comment]
The exit code when bad substitution happens (anywhere, not just in heredoc) is not correct yet.
Bad substitution does not set the exit code atm.

@LeaYeh What do you think, should the expander set the exit code, or whoever is calling the expander?

File: source/backend/redirection/heredoc.c
Comment (itislu) [inline source/backend/redirection/heredoc.c:1]: The return value of the expander can not only be `true` or `false`, but also `BAD_SUBSTITUTION`.
`BAD_SUBSTITUTION` should **abort** the heredoc, but not report a `HEREDOC_ERROR` (which should only happen for more critical error, like malloc error etc).

File: source/backend/redirection/heredoc_utils.c
Comment (itislu) [inline source/backend/redirection/heredoc_utils.c:1]: Move the new expansion part of the heredoc into heredoc_utils and replace the old `expand_heredoc_content()` function with it.

File: source/builtins/echo.c
Comment (itislu) [inline source/builtins/echo.c:1]: Use the same method as in the fix in string_utils.c: inrecement the string index variable with the return value of `ft_snprintf()`, which is the amount of characters written to the string.

File: source/utils/string_utils.c
Comment (itislu) [inline source/utils/string_utils.c:1]: The `delim` string needs to get added for every node in the `line_list`, even if the line is `NULL`.
Example:
```sh
cat << eof
foo
$NONEXIST
bar
eof
```
Correct result:
```sh
foo

bar
```
Wrong result:
```sh
foo
bar
```

File: source/utils/string_utils.c [hunk 1]
```diff
@@ -83,13 +82,10 @@ char	*concat_list_to_string(t_list *list, char *delim)
...
 		if (list->content)
-			ft_snprintf(&str[cur_len],
-				ft_strlen(list->content) + 1, "%s", list->content);
-		if (list->next && delim)
-		{
-			ft_snprintf(&str[cur_len], delim_len + 1, "%s", delim);
-			cur_len += delim_len;
-		}
+			cur_len += ft_snprintf(&str[cur_len],
+					ft_strlen(list->content) + 1, "%s", list->content);
```
Comment (itislu) [inline source/utils/string_utils.c:86]: This was a bug: `cur_len` was not incremented by the length of `list->content`.

File: include/defines.h [hunk 2]
```diff
@@ -61,7 +61,7 @@
 # define SUCCESS            0
 # define EXIT_SUCCESS       0
 # define GENERAL_ERROR      1
-# define BAD_SUBSTITUTION   2
+# define BAD_SUBSTITUTION   1
```
Comment (LeaYeh) [inline include/defines.h:64]: Could we reuse general error? I think they are all belong to it, we don't need to create a new one for each case

**Review by LeaYeh — COMMENTED** [review]
Too many error type point to same thing

File: include/defines.h (hunk 2 above)
Comment (itislu) [reply include/defines.h:64]: I understand what you mean. However, `BAD_SUBSTITUTION` and `GENERAL_ERROR` don't behave quite the same way.
Consider this setup:
`GENERAL_ERROR` example:
```sh
bash-5.1$ chmod 000 out
bash-5.1$ echo ABC > out
bash: out: Permission denied
bash-5.1$ echo $?
1
```
`BAD_SUBSTITUTION` example:
```sh
bash-5.1$ echo ${}
bash: ${}: bad substitution
bash-5.1$ echo $?
1
```
**Impact** of `GENERAL_ERROR`:
```sh
bash-5.1$ echo ABC > out || echo DEF
bash: out: Permission denied
DEF
```
**Impact** of `BAD_SUBSTITUTION`:
```sh
bash-5.1$ echo > out ${} || echo DEF
bash: ${}: bad substitution
```
* See #91

Comment (itislu) [reply include/defines.h:64]: But if you prefer I can create an enum for the return values of the expander and use it for internal error reporting only, like you did it with the `heredoc_status` enum.

Comment (itislu) [reply include/defines.h:64]: These are the exit codes defined in the bash source code:
```h
/* Values that can be returned by execute_command (). */
#define EXECUTION_FAILURE 1
#define EXECUTION_SUCCESS 0

/* Usage messages by builtins result in a return status of 2. */
#define EX_BADUSAGE	2

#define EX_MISCERROR	2

/* Special exit statuses used by the shell, internally and externally. */
#define EX_RETRYFAIL	124
#define EX_WEXPCOMSUB	125
#define EX_BINARY_FILE	126
#define EX_NOEXEC	126
#define EX_NOINPUT	126
#define EX_NOTFOUND	127

#define EX_SHERRBASE	256	/* all special error values are > this. */

#define EX_BADSYNTAX	257	/* shell syntax error */
#define EX_USAGE	258	/* syntax error in usage */
#define EX_REDIRFAIL	259	/* redirection failed */
#define EX_BADASSIGN	260	/* variable assignment error */
#define EX_EXPFAIL	261	/* word expansion failed */
#define EX_DISKFALLBACK	262	/* fall back to disk command from builtin */
```

Comment (LeaYeh) [reply include/defines.h:64]: Uhnnn interesting 🧐

**Review by LeaYeh — APPROVED** [review]

**Pull request merged**